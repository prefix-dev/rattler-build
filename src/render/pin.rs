use std::str::FromStr;

use rattler_conda_types::{MatchSpec, Version};
use serde::{de, Deserialize, Deserializer, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinExpression(#[serde(deserialize_with = "deserialize_pin_expression")] String);

fn deserialize_pin_expression<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    // A pin expression can only contain x and . (e.g. x.x.x)
    let s = String::deserialize(deserializer)?;
    let s = PinExpression::from_str(&s).map_err(de::Error::custom)?.0;
    Ok(s)
}

impl FromStr for PinExpression {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.chars().any(|c| c != 'x' && c != '.') {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "Invalid pin expression (can only contain x and .)",
            ));
        }
        Ok(PinExpression(s.to_string()))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pin {
    /// The name of the package to pin
    pub name: String,

    /// A pin to a version, using `x.x.x...` as syntax
    pub max_pin: Option<PinExpression>,

    /// A minimum pin to a version, using `x.x.x...` as syntax
    pub min_pin: Option<PinExpression>,

    /// If an exact pin is given, we pin the exact version & hash
    #[serde(default)]
    pub exact: bool,
}

impl Pin {
    pub fn apply(&self, version: &Version, hash: &str) -> Result<MatchSpec, anyhow::Error> {
        if self.exact {
            return Ok(MatchSpec::from_str(&format!(
                "{} {} {}",
                self.name, version, hash
            ))?);
        }
        let mut spec = self.name.clone();
        let version_str = version.to_string();

        // extract same amount of digits as the pin expression (in the form of x.x.x) from version str
        if let Some(min_pin) = &self.min_pin {
            // mumber of digits in pin expression
            let pin_digits = min_pin.0.chars().filter(|c| *c == 'x').count();
            // get version strin gup the to pin_digits dot
            let pin = version_str
                .split('.')
                .take(pin_digits)
                .collect::<Vec<_>>()
                .join(".");
            spec.push_str(&format!(" >={}", pin));
        }
        if let Some(max_pin) = &self.max_pin {
            // mumber of digits in pin expression
            let pin_digits = max_pin.0.chars().filter(|c| *c == 'x').count();
            // get version strin gup the to pin_digits dot
            let mut pin = version_str
                .split('.')
                .take(pin_digits)
                .map(|s| s.to_string())
                .collect::<Vec<_>>();

            // fill up with 0s
            while pin.len() < pin_digits {
                pin.push("0".to_string());
            }

            // increment last digit
            let last = pin.pop().unwrap().parse::<u64>().unwrap() + 1;
            pin.push(last.to_string());
            let pin = pin.join(".");

            if self.min_pin.is_some() {
                spec.push(',');
            } else {
                spec.push(' ');
            }
            spec.push_str(&format!("<{}", pin));
        }

        Ok(MatchSpec::from_str(&spec)?)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_apply_pin() {
        let pin = Pin {
            name: "foo".to_string(),
            max_pin: Some(PinExpression("x.x.x".to_string())),
            min_pin: Some(PinExpression("x.x.x".to_string())),
            exact: false,
        };

        let version = Version::from_str("1.2.3").unwrap();
        let hash = "1234567890";
        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3,<1.2.4");

        let short_version = Version::from_str("1").unwrap();
        let spec = pin.apply(&short_version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1,<1.0.1");

        let pin = Pin {
            name: "foo".to_string(),
            max_pin: Some(PinExpression("x.x.x".to_string())),
            min_pin: None,
            exact: false,
        };

        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo <1.2.4");

        let pin = Pin {
            name: "foo".to_string(),
            max_pin: None,
            min_pin: Some(PinExpression("x.x.x".to_string())),
            exact: false,
        };

        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3");
    }

    #[test]
    fn test_apply_exact_pin() {
        let pin = Pin {
            name: "foo".to_string(),
            max_pin: Some(PinExpression("x.x.x".to_string())),
            min_pin: Some(PinExpression("x.x.x".to_string())),
            exact: true,
        };

        let version = Version::from_str("1.2.3").unwrap();
        let hash = "h1234_0";
        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo ==1.2.3 h1234_0");
    }
}
