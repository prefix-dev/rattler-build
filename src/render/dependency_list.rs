use std::{fmt, str::FromStr};

use super::pin::Pin;
use rattler_conda_types::MatchSpec;
use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer, Serialize,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinSubpackage {
    pub pin_subpackage: Pin,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Compiler {
    pub compiler: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(untagged)]
pub enum Dependency {
    #[serde(deserialize_with = "deserialize_match_spec")]
    Spec(MatchSpec),
    PinSubpackage(PinSubpackage),
    Compiler(Compiler),
}

impl<'de> Deserialize<'de> for Dependency {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct DependencyVisitor;

        impl<'de> Visitor<'de> for DependencyVisitor {
            type Value = Dependency;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str(
                    "a string starting with '__COMPILER', '__PIN_SUBPACKAGE', or a MatchSpec",
                )
            }

            fn visit_str<E>(self, value: &str) -> Result<Dependency, E>
            where
                E: de::Error,
            {
                if let Some(compiler) = value.strip_prefix("__COMPILER ") {
                    Ok(Dependency::Compiler(Compiler {
                        compiler: compiler.to_lowercase(),
                    }))
                } else if let Some(pin) = value.strip_prefix("__PIN_SUBPACKAGE ") {
                    Ok(Dependency::PinSubpackage(PinSubpackage {
                        pin_subpackage: Pin::from_internal_repr(pin),
                    }))
                } else {
                    // Assuming MatchSpec can be constructed from a string.
                    MatchSpec::from_str(value)
                        .map(Dependency::Spec)
                        .map_err(de::Error::custom)
                }
            }
        }

        deserializer.deserialize_str(DependencyVisitor)
    }
}

pub type DependencyList = Vec<Dependency>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_matchspecs() {
        let spec = r#"
- python
- python >=3.6
- python >=3.6,<3.7
- python >=3.6,<3.7.0a0
- python >=3.6,<3.7.0a0[build=py36h6de7cb9_0]
        "#
        .trim();
        let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        insta::assert_yaml_snapshot!(spec);
    }

    #[test]
    fn test_render_compiler() {
        // The following works, but doesn't play nicely with Jinja (since the jinja statements)
        // are quoted and can't produce proper YAML easily ... so we use the __COMPILER
        // syntax instead. We could / should revisit this!
        // let spec = r#"
        // - { compiler: "c" }
        // - { compiler: "cxx" }
        // - { compiler: "fortran" }
        // - { compiler: "rust" }
        //         "#
        //         .trim();
        //         let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        //         insta::assert_yaml_snapshot!(spec);
        let spec = r#"
        - __COMPILER C
        - __COMPILER CXX
        - __COMPILER FORTRAN
        - __COMPILER RUST
        "#;
        let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        insta::assert_yaml_snapshot!(spec);
    }

    #[test]
    fn test_render_pin_subpackage() {
        let pin = "- __PIN_SUBPACKAGE name MAX_PIN= MIN_PIN=x.x.x EXACT=False";
        let spec: DependencyList = serde_yaml::from_str(pin).unwrap();
        insta::assert_yaml_snapshot!(spec);

        let pin = "- __PIN_SUBPACKAGE super-package MAX_PIN=x.x MIN_PIN=x.x.x EXACT=true";
        let spec: DependencyList = serde_yaml::from_str(pin).unwrap();
        let p = &spec[0];
        let p = match p {
            Dependency::PinSubpackage(p) => p,
            _ => panic!("Expected PinSubpackage"),
        };
        assert_eq!(p.pin_subpackage.name.as_normalized(), "super-package");
        assert_eq!(
            p.pin_subpackage.max_pin.as_ref().unwrap().to_string(),
            "x.x"
        );
        assert_eq!(
            p.pin_subpackage.min_pin.as_ref().unwrap().to_string(),
            "x.x.x"
        );
        assert!(p.pin_subpackage.exact);
    }
}
