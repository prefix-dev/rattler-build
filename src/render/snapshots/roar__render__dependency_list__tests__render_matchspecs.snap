---
source: src/render/dependency_list.rs
expression: spec
---
- name: python
- name: python
  version: ">=3.6"
- name: python
  version: ">=3.6,<3.7"
- name: python
  version: ">=3.6,<3.7.0a0"
- name: python
  version: ">=3.6,<3.7.0a0"
  build: py36h6de7cb9_0

